src/transaction.rs

```diff
@@ -0,0 +1,119 @@
+// This Source Code Form is subject to the terms of the Mozilla Public
+// License, v. 2.0. If a copy of the MPL was not distributed with this
+// file, You can obtain one at http://mozilla.org/MPL/2.0/.
+//
+// Copyright (c) DUSK NETWORK. All rights reserved.
+
+//! Verifier data for the transfer circuits
+
+use alloc::string::String;
+use alloc::vec::Vec;
+
+#[cfg(feature = "rkyv-impl")]
+use rkyv::{Archive, Deserialize, Serialize};
+
+use dusk_bls12_381::BlsScalar;
+use dusk_bytes::Serializable;
+use dusk_poseidon::cipher::PoseidonCipher;
+
+use crate::Crossover;
+use crate::Fee;
+use crate::Message;
+use crate::Note;
+
+/// For the purposes of our transaction model, ModuleId is always a BlsScalar
+pub type ModuleId = BlsScalar;
+
+const STCO_MESSAGE_SIZE: usize = 7 + 2 * PoseidonCipher::cipher_size();
+const STCT_MESSAGE_SIZE: usize = 5 + PoseidonCipher::cipher_size();
+
+/// A phoenix transaction.
+#[derive(Debug, Clone)]
+#[cfg_attr(
+    feature = "rkyv-impl",
+    derive(Archive, Serialize, Deserialize),
+    archive_attr(derive(bytecheck::CheckBytes))
+)]
+pub struct Transaction {
+    /// The root of the transfer tree on top of which this transaction is
+    /// based.
+    pub anchor: BlsScalar,
+    /// The nullifiers of the notes this transaction spends.
+    pub nullifiers: Vec<BlsScalar>,
+    /// The output notes of this transaction.
+    pub outputs: Vec<Note>,
+    /// Describes the fee to be paid for this transaction.
+    pub fee: Fee,
+    /// A crossover is used to transferring funds to a contract - i.e. in
+    /// [`Stct`] and [`Stco`].
+    pub crossover: Option<Crossover>,
+    /// Serialized proof of the `Execute` circuit for this transaction.
+    pub proof: Vec<u8>,
+    /// A call to a contract. The `Vec<u8>` must be an `rkyv`ed representation
+    /// of the data the contract expects, and the `String` the name of the
+    /// function to call.
+    pub call: Option<(ModuleId, String, Vec<u8>)>,
+}
+
+impl Transaction {
+    /// Return input bytes to a hash function for the transaction.
+    #[must_use]
+    pub fn to_hash_input_bytes(&self) -> Vec<u8> {
```

Comment:
It might be wise to introduce another function that computes the hash input bytes from the decomposed inputs. This is because the `tx_hash` has to be computed before a transaction can be properly created.

```rust
pub fn hash_input_from_components(
        nullifiers: &[BlsScalar],
        outputs: &[Note],
        anchor: &BlsScalar,
        fee: &Fee,
        crossover: &Option<Crossover>,
        call: &Option<(ModuleId, String, Vec<u8>)>,
    ) {
// ...
}
```
The current function would then call this one.

Reply:
Also, there's the issue of serialization of the transaction. We should probably have a `to_var_bytes` function that serializes a transaction, and a `from_slice` or `from_bytes` that deserializes it.